text/unify-theorem-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,193 @@
+    providing only its body.
+
+  - In the upcoming refactoring of the reference manual, we would like
+    the first chapter to be about the core language that the kernel
+    understands.  We couldn't give examples using `Theorem` because
+    this command always open the proof mode.  Yet, it is important to
+    be able to explain that the kernel can define opaque
+    constants, i.e. constants with a body which behave like axioms.
+
+  - It would allow deprecating the akward `Proof @term` syntax.
+
+- Fixing the inconsistent semantics of `Theorem` and `Theorem
+  ... with`.  The first syntax form behaves like `Definition` but the
+  second one behaves like `Fixpoint`.
+
+  A previous attempted solution was simply to deprecate the `Theorem
+  ... with` syntax in favor of `Fixpoint ... with` in proof mode
+  ([coq/coq#8265](https://github.com/coq/coq/pull/8265)).  However,
+  this syntax is used in CompCert and Xavier Leroy argued in favor of
+  keeping it, in particular on the basis that people who read and
+  review formalizations might appreciate the distinction of intent
+  between a `Theorem` and a `Fixpoint`.  Still, there is no reason why
+  we should explicitly distinguish between a `Definition` and a
+  `Fixpoint` but not between a non-recursive and a recursive
+  `Theorem`.
+
+  The proposed solution is to keep the existing syntax `Theorem
+  ... with` but encourage that it comes with an explicit `recursive`
+  annotation (in the form of an attribute).  This also opens the way
+  to supporting (non-mutually) recursive `#[ recursive ] Theorem`, and
+  `#[ recursive ] Definition` as an equivalent to `Fixpoint` to define
+  (possibly mutually) recursive objects.
+
+# Detailed design
+
+The current syntax is:
+
+- for `Definition` / `Example`:
+  ```
+  {| Definition | Example } @ident_decl @def_body
+
+  def_body ::= {* @binder } {? : @type } := {? @reduce } @term
+             | {* @binder } : @type
+  ```
+
+- for `Fixpoint`:
+  ```
+  Fixpoint @fix_def {* with @fix_def }
+
+  fix_def ::= @ident_decl {* @binder } {? @fixannot } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `CoFixpoint`:
+  ```
+  CoFixpoint @cofix_def {* with @cofix_def }
+
+  cofix_def ::= @ident_decl {* @binder } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `Theorem` and variants:
+  ```
+  @thm_token @ident_decl {* @binder } : @type {* with @ident_decl {* @binder } : @type }
+
+  thm_token ::= Theorem
+              | Lemma
+              | Fact
+              | Remark
+              | Corollary
+              | Proposition
+              | Property
+  ```
+
+The new syntax would be:
+
+```
+@decl_token @decl_def {* with @decl_def }
+
+decl_token := Definition
+            | Example
+            | Fixpoint
+            | CoFixpoint
+            | Theorem
+            | Lemma
+            | Fact
+            | Remark
+            | Corollary
+            | Proposition
+            | Property
```

Comment:
What about `Instance`? It seems `Instance` should be a synonym *except* for instance registration. Instead, `Instance` supports a random different subset of features. In particular, mutually recursive instances currently give a parse error (https://github.com/coq/coq/issues/7913).
Hopefully the difficulty of implementation is a separate concern.

Reply:
Indeed, `Instance` should be merged as well and be the same as `Definition` with an instance registration attribute.

Reply:
Note that unlike every other definition-like command, `Instance` supports `:= { ... }` syntax for using bidi typechecking for record bodies/types.  (I hear this syntax is going to be deprecated someday?)

Reply:
I was aware of this syntax but not of what it did. @maximedenes recently introduced some generic bidi typechecking features (in the `Arguments` command). Are they enough to replace this syntax? Should we also provide an attribute for this?

Reply:
I am not sure if `Arguments` is enough.  I think it might be almost enough, if you have an attribute that automatically sets it locally at the point in the record constructor between the last parameter and the first field.  However, I think there's also some magic where you can use the unqualified version of the field names even if the module the record is in isn't imported, or something like that (I think similar to how you don't always need to qualify constructor names in OCaml matches).  This bit of it seems more fine to deprecate, though, if need be.

Note that this only makes sense for definitions / instances constructing records.

Reply:
Also, another difference: `Instance` does not require a name (much like `Goal`), while all the other definition-like commands do.  That is, `Existing Class True.  Instance : True := I.` is valid, and will generate an automatic name for the instance.

Reply:
> Should we also provide an attribute for this?

I would not object, and it seems like it might be useful.  Something even more useful would be `Set Automatic Bidi Constructors` or similar that made it so that all inductives declared while this flag is set have automatic bidi declarations for all of their constructors, which place the bidi location between the last (uniform?) inductive parameter and the first constructor argument.  I'll go report a feature request for this.

Reply:
https://github.com/coq/coq/issues/11702

Reply:
> However, I think there's also some magic where you can use the unqualified version of the field names even if the module the record is in isn't imported, or something like that (I think similar to how you don't always need to qualify constructor names in OCaml matches). This bit of it seems more fine to deprecate, though, if need be.

I’d rather we make it more widely available — field names should always be disambiguated by type even when not imported…

Reply:
@Blaisorblade Could you open another issue for this? It should probably be tracked outside this CEP.